File: pxpoint/PxPointCorrelation.cs
using System;
using System.Threading;

namespace PxPoint.Correlation
{
    /// <summary>
    /// C# version of the PxPoint correlation system for multi-process pipeline correlation
    /// Supports hierarchical correlation IDs: Pipeline -> Process -> Activity
    /// </summary>
    public class PxPointCorrelationManager
    {
        private static readonly Lazy<PxPointCorrelationManager> _instance = 
            new Lazy<PxPointCorrelationManager>(() => new PxPointCorrelationManager());
        
        public static PxPointCorrelationManager Instance => _instance.Value;
        
        // Thread-local storage for activity-level correlation
        private static readonly ThreadLocal<string> ThreadActivityId = 
            new ThreadLocal<string>(() => string.Empty);
        
        // Process-level correlation (shared across threads in same process)
        private string _pipelineId = string.Empty;
        private string _processId = string.Empty;
        
        private PxPointCorrelationManager() { }
        
        // Pipeline-level correlation (shared across all processes)
        public void SetPipelineId(string pipelineId)
        {
            _pipelineId = pipelineId ?? string.Empty;
        }
        
        public string GetPipelineId()
        {
            return _pipelineId;
        }
        
        // Process-level correlation (unique per process type)
        public void SetProcessId(string processId)
        {
            _processId = processId ?? string.Empty;
        }
        
        public string GetProcessId()
        {
            return _processId;
        }
        
        // Activity-level correlation (within process activities)
        public void SetActivityId(string activityId)
        {
            ThreadActivityId.Value = activityId ?? string.Empty;
        }
        
        public string GetActivityId()
        {
            return ThreadActivityId.Value ?? string.Empty;
        }
        
        public void ClearActivityId()
        {
            ThreadActivityId.Value = string.Empty;
        }
        
        // Generate new correlation IDs
        public string GeneratePipelineId()
        {
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var uuid = Guid.NewGuid().ToString("N")[..8];
            return $"pxp-{timestamp}-{uuid}";
        }
        
        public string GenerateProcessId(string processType)
        {
            if (string.IsNullOrEmpty(_pipelineId))
            {
                _pipelineId = GeneratePipelineId();
            }
            
            var uuid = Guid.NewGuid().ToString("N")[..8];
            return $"{_pipelineId}-{processType}-{uuid}";
        }
        
        public string GenerateActivityId(string activityName)
        {
            var baseId = string.IsNullOrEmpty(_processId) ? string.Empty : _processId;
            var uuid = Guid.NewGuid().ToString("N")[..8];
            
            return string.IsNullOrEmpty(baseId) 
                ? $"{activityName}-{uuid}"
                : $"{baseId}-{activityName}-{uuid}";
        }
        
        // Environment variable integration for cross-process correlation
        public void LoadFromEnvironment()
        {
            var pipelineEnv = Environment.GetEnvironmentVariable("PXPOINT_PIPELINE_ID");
            if (!string.IsNullOrEmpty(pipelineEnv))
            {
                _pipelineId = pipelineEnv;
            }
            
            var processEnv = Environment.GetEnvironmentVariable("PXPOINT_PROCESS_ID");
            if (!string.IsNullOrEmpty(processEnv))
            {
                _processId = processEnv;
            }
        }
        
        public void SaveToEnvironment()
        {
            if (!string.IsNullOrEmpty(_pipelineId))
            {
                Environment.SetEnvironmentVariable("PXPOINT_PIPELINE_ID", _pipelineId);
            }
            
            if (!string.IsNullOrEmpty(_processId))
            {
                Environment.SetEnvironmentVariable("PXPOINT_PROCESS_ID", _processId);
            }
        }
        
        // Get full correlation context for logging
        public string GetFullCorrelationId()
        {
            var parts = new List<string>();
            
            if (!string.IsNullOrEmpty(_pipelineId))
            {
                parts.Add($"pipeline:{_pipelineId}");
            }
            
            if (!string.IsNullOrEmpty(_processId))
            {
                parts.Add($"process:{_processId}");
            }
            
            var activityId = GetActivityId();
            if (!string.IsNullOrEmpty(activityId))
            {
                parts.Add($"activity:{activityId}");
            }
            
            return string.Join("|", parts);
        }
    }
    
    /// <summary>
    /// RAII scope for activity-level correlation
    /// </summary>
    public class ActivityCorrelationScope : IDisposable
    {
        private readonly string _previousId;
        private bool _disposed = false;
        
        public ActivityCorrelationScope(string activityName)
        {
            var manager = PxPointCorrelationManager.Instance;
            _previousId = manager.GetActivityId();
            var newActivityId = manager.GenerateActivityId(activityName);
            manager.SetActivityId(newActivityId);
        }
        
        public void Dispose()
        {
            if (!_disposed)
            {
                var manager = PxPointCorrelationManager.Instance;
                if (string.IsNullOrEmpty(_previousId))
                {
                    manager.ClearActivityId();
                }
                else
                {
                    manager.SetActivityId(_previousId);
                }
                _disposed = true;
            }
        }
    }
    
    /// <summary>
    /// RAII scope for process initialization
    /// </summary>
    public class ProcessCorrelationScope : IDisposable
    {
        private readonly bool _createdNewPipeline;
        private bool _disposed = false;
        
        public ProcessCorrelationScope(string processType)
        {
            var manager = PxPointCorrelationManager.Instance;
            
            // Try to load existing correlation from environment
            manager.LoadFromEnvironment();
            
            // If no pipeline ID exists, create a new one
            if (string.IsNullOrEmpty(manager.GetPipelineId()))
            {
                manager.SetPipelineId(manager.GeneratePipelineId());
                _createdNewPipeline = true;
            }
            else
            {
                _createdNewPipeline = false;
            }
            
            // Generate process ID for this process type
            var processId = manager.GenerateProcessId(processType);
            manager.SetProcessId(processId);
            
            // Save to environment for child processes
            manager.SaveToEnvironment();
        }
        
        public void Dispose()
        {
            if (!_disposed)
            {
                // Clean up if we created the pipeline
                if (_createdNewPipeline)
                {
                    Environment.SetEnvironmentVariable("PXPOINT_PIPELINE_ID", null);
                    Environment.SetEnvironmentVariable("PXPOINT_PROCESS_ID", null);
                }
                _disposed = true;
            }
        }
    }
}

File: pxpoint/pxpoint_correlation.cpp
#include "pxpoint_correlation.h"
#include <random>
#include <sstream>
#include <iomanip>
#include <chrono>

namespace pxpoint {

thread_local std::string PxPointCorrelationManager::thread_activity_id_;

PxPointCorrelationManager& PxPointCorrelationManager::getInstance() {
    static PxPointCorrelationManager instance;
    return instance;
}

void PxPointCorrelationManager::setPipelineId(const std::string& pipeline_id) {
    pipeline_id_ = pipeline_id;
}

std::string PxPointCorrelationManager::getPipelineId() const {
    return pipeline_id_;
}

void PxPointCorrelationManager::setProcessId(const std::string& process_id) {
    process_id_ = process_id;
}

std::string PxPointCorrelationManager::getProcessId() const {
    return process_id_;
}

void PxPointCorrelationManager::setActivityId(const std::string& activity_id) {
    thread_activity_id_ = activity_id;
}

std::string PxPointCorrelationManager::getActivityId() const {
    return thread_activity_id_;
}

void PxPointCorrelationManager::clearActivityId() {
    thread_activity_id_.clear();
}

std::string PxPointCorrelationManager::generateUuid() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);
    
    std::stringstream ss;
    ss << std::hex;
    for (int i = 0; i < 8; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 4; i++) ss << dis(gen);
    ss << "-4"; // Version 4 UUID
    for (int i = 0; i < 3; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 4; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 12; i++) ss << dis(gen);
    
    return ss.str();
}

std::string PxPointCorrelationManager::generatePipelineId() {
    auto now = std::chrono::system_clock::now();
    auto time_t = std::chrono::system_clock::to_time_t(now);
    
    std::stringstream ss;
    ss << "pxp-" << time_t << "-" << generateUuid().substr(0, 8);
    return ss.str();
}

std::string PxPointCorrelationManager::generateProcessId(const std::string& process_type) {
    if (pipeline_id_.empty()) {
        pipeline_id_ = generatePipelineId();
    }
    
    std::stringstream ss;
    ss << pipeline_id_ << "-" << process_type << "-" << generateUuid().substr(0, 8);
    return ss.str();
}

std::string PxPointCorrelationManager::generateActivityId(const std::string& activity_name) {
    if (process_id_.empty()) {
        return activity_name + "-" + generateUuid().substr(0, 8);
    }
    
    std::stringstream ss;
    ss << process_id_ << "-" << activity_name << "-" << generateUuid().substr(0, 8);
    return ss.str();
}

void PxPointCorrelationManager::loadFromEnvironment() {
    const char* pipeline_env = std::getenv("PXPOINT_PIPELINE_ID");
    if (pipeline_env) {
        pipeline_id_ = pipeline_env;
    }
    
    const char* process_env = std::getenv("PXPOINT_PROCESS_ID");
    if (process_env) {
        process_id_ = process_env;
    }
}

void PxPointCorrelationManager::saveToEnvironment() {
    if (!pipeline_id_.empty()) {
        setenv("PXPOINT_PIPELINE_ID", pipeline_id_.c_str(), 1);
    }
    
    if (!process_id_.empty()) {
        setenv("PXPOINT_PROCESS_ID", process_id_.c_str(), 1);
    }
}

std::string PxPointCorrelationManager::getFullCorrelationId() const {
    std::stringstream ss;
    
    if (!pipeline_id_.empty()) {
        ss << "pipeline:" << pipeline_id_;
    }
    
    if (!process_id_.empty()) {
        if (ss.tellp() > 0) ss << "|";
        ss << "process:" << process_id_;
    }
    
    if (!thread_activity_id_.empty()) {
        if (ss.tellp() > 0) ss << "|";
        ss << "activity:" << thread_activity_id_;
    }
    
    return ss.str();
}

// ActivityCorrelationScope implementation
ActivityCorrelationScope::ActivityCorrelationScope(const std::string& activity_name) {
    auto& manager = PxPointCorrelationManager::getInstance();
    previous_id_ = manager.getActivityId();
    std::string new_activity_id = manager.generateActivityId(activity_name);
    manager.setActivityId(new_activity_id);
}

ActivityCorrelationScope::~ActivityCorrelationScope() {
    auto& manager = PxPointCorrelationManager::getInstance();
    if (previous_id_.empty()) {
        manager.clearActivityId();
    } else {
        manager.setActivityId(previous_id_);
    }
}

// ProcessCorrelationScope implementation
ProcessCorrelationScope::ProcessCorrelationScope(const std::string& process_type) 
    : created_new_pipeline_(false) {
    auto& manager = PxPointCorrelationManager::getInstance();
    
    // Try to load existing correlation from environment
    manager.loadFromEnvironment();
    
    // If no pipeline ID exists, create a new one
    if (manager.getPipelineId().empty()) {
        manager.setPipelineId(manager.generatePipelineId());
        created_new_pipeline_ = true;
    }
    
    // Generate process ID for this process type
    std::string process_id = manager.generateProcessId(process_type);
    manager.setProcessId(process_id);
    
    // Save to environment for child processes
    manager.saveToEnvironment();
}

ProcessCorrelationScope::~ProcessCorrelationScope() {
    // Clean up if we created the pipeline
    if (created_new_pipeline_) {
        unsetenv("PXPOINT_PIPELINE_ID");
        unsetenv("PXPOINT_PROCESS_ID");
    }
}

} // namespace pxpoint